fix #4332: setting ZSTD_NBTHREADS=0 via environment variable

Cyan4973 opened the pull request:
Since `v1.5.7`, setting `ZSTD_NBTHREADS=0` via environment variable doesn't work, due to a logic mistake in the initialization routine.
Reported in #4332, by [JcKlomp](https://github.com/JcKlomp).

Cyan4973 committed:
fix #4332: setting ZSTD_NBTHREADS=0 via environment variables

Cyan4973 committed:
removed useless assert()

terrelln commented on tests/cli-tests/compression/multi-threaded.sh:
```diff
@@ -10,7 +10,3 @@ zstd -T0 -f file -q                     ; zstd -t file.zst
 zstd -T0 --auto-threads=logical -f file -q ; zstd -t file.zst
 zstd -T0 --auto-threads=physical -f file -q ; zstd -t file.zst
 zstd -T0 --jobsize=1M -f file -q        ; zstd -t file.zst
```
Should we add a test here that tests the environment variable works?

If we print the number of threads under some verbosity we could do something like:

```
ZSTD_NBTHREADS=0 zstd -f file -vv 2>&1 | grep "physical core(s) detected"
zstd -tq file.zst
ZSTD_NBTHREADS=4 zstd -f file -vv 2>&1 | grep "4 threads"
zstd -tq file.zst
```

Cyan4973 replied:
Yes, definitely,
I was considering to do it in another PR, in order to observe the test failure under current conditions, i.e. without the patch.
But I could also just add it to this PR as a follow up commit.

terrelln commented on programs/zstdcli.c:
```diff
@@ -1332,27 +1335,16 @@ int main(int argCount, const char* argv[])
     DISPLAYLEVEL(3, WELCOME_MESSAGE);
 
 #ifdef ZSTD_MULTITHREAD
-    if ((operation==zom_decompress) && (nbWorkers > 1)) {
-        DISPLAYLEVEL(2, "Warning : decompression does not support multi-threading\n");
-    }
```
This was added due to Issue #2918. This may be a useful warning to keep.

Cyan4973 replied:
Issue is, with current code,
there is nothing to tell that `nbWorkers` wasn't set.
Whether it's set via the environment variable, or via the core count detection, it's necessarily set to "some value".
There is no difference with a user setting this value manually via command line.

Therefore, on any modern system with >= 8 logical cores, `nbWorkers >= 2`, and consequently _always_ triggers this warning.

This could be solved by adding yet another variable, which would track if `nbWorkers` was set explicitly, or if it happens to be set to a value `>1` just as a consequence of automatic core count detection. We would also have to decide if setting `nbWorkers` via environment variable count or not to trigger this warning.

So it's a bit more work, a bit more questions, and specific to the topic of displaying a warning during decompression mode. It felt it could deserve its own PR for proper focus.

Cyan4973 committed:
add test

checks that ZSTD_NBTHREADS triggers the expected verbose message

Also: checked that the new test script fails on current `dev` branch, and is fixed by this branch

Cyan4973 commented:
test added in this PR,
checked manually on `dev` branch (where it produces an error signal, as expected).

Cyan4973 committed:
fix FreeBSD

use an alias instead of a function

also: added more traces and updated version nb to v1.5.8

Cyan4973 commented:
In the process, fixed a latent limitation of the test shell script `playTests.sh`,
where was unable to pass environment variables (such as `ZSTD_NBTHREADS`) to `zstd`
specifically in FreeBSD environment (and possibly other *BSD).

Cyan4973 committed:
fixed ShellCheck warning
